x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TransportSqlQueryAction.java

```diff
@@ -64,28 +73,74 @@ public TransportSqlQueryAction(Settings settings, ClusterService clusterService,
         sqlLicenseChecker.checkIfSqlAllowed(request.mode());
-        operation(planExecutor, request, listener, username(securityContext), clusterName(clusterService));
+        operation(planExecutor, request, listener, username(securityContext), clusterName(clusterService), transportService,
+            clusterService);
     }
 
     /**
      * Actual implementation of the action. Statically available to support embedded mode.
      */
     static void operation(PlanExecutor planExecutor, SqlQueryRequest request, ActionListener<SqlQueryResponse> listener,
-                                 String username, String clusterName) {
+                                 String username, String clusterName, TransportService transportService, ClusterService clusterService) {
         // The configuration is always created however when dealing with the next page, only the timeouts are relevant
         // the rest having default values (since the query is already created)
         SqlConfiguration cfg = new SqlConfiguration(request.zoneId(), request.fetchSize(), request.requestTimeout(), request.pageTimeout(),
                 request.filter(), request.mode(), request.clientId(), request.version(), username, clusterName,
                 request.fieldMultiValueLeniency(), request.indexIncludeFrozen());
 
         if (Strings.hasText(request.cursor()) == false) {
+            Holder<Boolean> retrySecondTime = new Holder<Boolean>(false);
             planExecutor.sql(cfg, request.query(), request.params(),
+                wrap(p -> listener.onResponse(createResponseWithSchema(request, p)), e -> {
+                    // the search request will likely run on nodes with different versions of ES
+                    // we will retry on a node with an older version that should generate a backwards compatible _search request
+                    if (e instanceof SearchPhaseExecutionException
+                        && ((SearchPhaseExecutionException) e).getCause() instanceof VersionMismatchException) {
+
+                        SearchPhaseExecutionException spee = (SearchPhaseExecutionException) e;
+                        if (log.isTraceEnabled()) {
+                            log.trace("Caught exception type [{}] with cause [{}].", e.getClass().getName(), e.getCause());
+                        }
+                        DiscoveryNode localNode = clusterService.state().nodes().getLocalNode();
+                        DiscoveryNode candidateNode = null;
+                        for (DiscoveryNode node : clusterService.state().nodes()) {
+                            // find the first node that's older than the current node
+                            if (node != localNode && node.getVersion().before(localNode.getVersion())) {
+                                candidateNode = node;
+                                break;
+                            }
+                        }
+                        if (candidateNode != null) {
```

Comment:
This class and the one in EQL can be simplified by moving the common code into a base class in QL.
The exception check plus `DiscoveryNode` selection, plus logging and retry are the same.
The only differences that I can see are calling transportService and the planExecutor call which can be passed as `Runnable` or `Function` in case a property needs to passed in.

Reply:
I've moved the common code to a separate method in QL.